### DISCUSSION TITLE
Neoclip: multi-platform clipboard provider w/o extra dependencies

### DISCUSSION THREAD

**Discussion by matveyt** [neovim/neovim#30534] [General]
This is a request for comments for my project [Neoclip](https://github.com/matveyt/neoclip).

As we all know, Neovim had dropped all system-specific code to access clipboard. While it was a flexible design, but it led to several common problems the users seem to encounter all the time: need for extra utlities or specific GUIs; poor performance under some circumstances; unreliable detection of Visual-Blocks and such. There were some attempts to resolve this problem, like [this](https://github.com/bfredl/nvim-x11) or [this](https://github.com/neovim/neovim/pull/12452), but none is complete and ready for use.

So I present my solution: it is a plugin and several binary extension modules for major platforms providing uniform API to access system clipboard. The project state is as follows:

+ multi-platform: Windows, macOS, X11 and Wayland;
+ pure plugin, no patches to Neovim needed;
+ only essential dependencies, such as wayland-client; no special utilities, GUIs, fat toolkits etc.
+ Vim-compatible custom clipboard formats for all platforms; able to detect Visual-Blocks;
+ X11: support for rare features like INCR; can paste >256K from Gtk apps;
- No binaries at the moment, so user must build from source
- Not enough testing and user experience;
- Wayland: [wlr-data-control](https://wayland.app/protocols/wlr-data-control-unstable-v1) protocol is required; if it's not available (for example, GNOME/Mutter) Neoclip falls back to X11 module;

So I'd be glad to get any feedback from all of you. If you think it's a good addition to your plugin collection; or maybe it's even worth shipping Neoclip as default plugin with Neovim runtime; or maybe you can spot some bugs and so on.

Thank you!

**Comment by bfredl** [comment]
Interesting work. We could avoid the threading of X11 by using the fast_events uv loop. this could be accessed via lua, though we could expose the uv handle to C code directly.

**Comment by matveyt** [comment]
> We could avoid the threading of X11 by using the fast_events uv loop.

I don't see how this could be done from inside a shared object, unless something would've been exported to Lua global namespace.

**Comment by justinmk** [comment]
@matveyt has neoclip found its way into homebrew or similar? Any plans there?

**Reply by shofel** [reply]
@justinmk @matveyt I prepared a homebrew formula https://github.com/matveyt/neoclip/pull/8,
and a nix flake https://github.com/matveyt/neoclip/pull/7/files

Please check if this can be helpful

**Comment by matveyt** [comment]
@justinmk Not until now. But I'll take a look.

**Comment by feoh** [comment]
Shot in the dark: Has anyone tested this under WSL2 with the standard Ubuntu distro?

I had to struggle a bit to keep Smartyank from detecting and using WSL2's imperfect Wayland support and instead use the terminal's OSC support.

I can try it myself the day after tomorrow. I'll be working on a Mac laptop until then.
